diff --git a/includes/core-phpunit/includes/abstract-testcase.php b/includes/core-phpunit/includes/abstract-testcase.php
index f2978644..e092beca 100644
--- a/includes/core-phpunit/includes/abstract-testcase.php
+++ b/includes/core-phpunit/includes/abstract-testcase.php
@@ -20,6 +20,8 @@ abstract class WP_UnitTestCase_Base extends PHPUnit_Adapter_TestCase {
 	protected $expected_doing_it_wrong = array();
 	protected $caught_doing_it_wrong   = array();
 
+    private static ?string $calledClass = null;
+
 	protected static $hooks_saved = array();
 	protected static $ignore_files;
 
@@ -37,7 +39,7 @@ abstract class WP_UnitTestCase_Base extends PHPUnit_Adapter_TestCase {
 	 *
 	 * @return WP_UnitTest_Factory The fixture factory.
 	 */
-	protected static function factory() {
+	public static function factory() {
 		static $factory = null;
 		if ( ! $factory ) {
 			$factory = new WP_UnitTest_Factory();
@@ -53,7 +55,7 @@ protected static function factory() {
 	 * @return string The class name.
 	 */
 	public static function get_called_class() {
-		return get_called_class();
+		return self::$called_class ?? get_called_class();
 	}
 
 	/**
@@ -66,10 +68,12 @@ public static function set_up_before_class() {
 
 		$wpdb->suppress_errors = false;
 		$wpdb->show_errors     = true;
-		$wpdb->db_connect();
-		ini_set( 'display_errors', 1 );
+        if ( empty( lucatume\WPBrowser\Utils\Property::readPrivate( $wpdb, 'dbh' ) ) ) {
+            self::fail( 'The database connection went away. A `setUpBeforeClassMethod` likely closed the connection.' );
+        }
+        ini_set( 'display_errors', 1 );
 
-		$class = get_called_class();
+		$class = self::$calledClass ?? get_called_class();
 
 		if ( method_exists( $class, 'wpSetUpBeforeClass' ) ) {
 			call_user_func( array( $class, 'wpSetUpBeforeClass' ), static::factory() );
@@ -82,7 +86,7 @@ public static function set_up_before_class() {
 	 * Runs the routine after all tests have been run.
 	 */
 	public static function tear_down_after_class() {
-		$class = get_called_class();
+        $class = self::$calledClass ?? get_called_class();
 
 		if ( method_exists( $class, 'wpTearDownAfterClass' ) ) {
 			call_user_func( array( $class, 'wpTearDownAfterClass' ) );
@@ -651,7 +655,7 @@ public function expectedDeprecated() {
 	 *
 	 * @since 4.2.0
 	 */
-	protected function assert_post_conditions() {
+	public function assert_post_conditions() {
 		$this->expectedDeprecated();
 	}
 
@@ -1660,4 +1664,9 @@ public static function touch( $file ) {
 
 		touch( $file );
 	}
+
+	public function setCalledClass(string $class): void
+	{
+		self::$calledClass = $class;
+	}
 }
