diff --git a/includes/core-phpunit/includes/testcase-rest-api.php b/includes/core-phpunit/includes/testcase-rest-api.php
index 27349809..2107abfc 100644
--- a/includes/core-phpunit/includes/testcase-rest-api.php
+++ b/includes/core-phpunit/includes/testcase-rest-api.php
@@ -1,6 +1,8 @@
 <?php
 
-abstract class WP_Test_REST_TestCase extends WP_UnitTestCase {
+namespace lucatume\WPBrowser\TestCase;
+
+abstract class WP_Test_REST_TestCase extends WPTestCase {
 
 	/**
 	 * Asserts that the REST API response has the specified error.
