diff --git a/includes/core-phpunit/includes/testcase-rest-api.php b/includes/core-phpunit/includes/testcase-rest-api.php
index 99dd5e30..e9005563 100644
--- a/includes/core-phpunit/includes/testcase-rest-api.php
+++ b/includes/core-phpunit/includes/testcase-rest-api.php
@@ -1,6 +1,8 @@
 <?php
 
-abstract class WP_Test_REST_TestCase extends WP_UnitTestCase {
+namespace lucatume\WPBrowser\TestCase;
+
+abstract class WPRestApiTestCase extends WPTestCase {
 
 	/**
 	 * Asserts that the REST API response has the specified error.
@@ -8,10 +10,10 @@ abstract class WP_Test_REST_TestCase extends WP_UnitTestCase {
 	 * @since 4.4.0
 	 * @since 6.6.0 Added the `$message` parameter.
 	 *
-	 * @param string|int                $code     Expected error code.
-	 * @param WP_REST_Response|WP_Error $response REST API response.
-	 * @param int                       $status   Optional. Status code.
-	 * @param string                    $message  Optional. Message to display when the assertion fails.
+	 * @param string|int                  $code     Expected error code.
+	 * @param \WP_REST_Response|\WP_Error $response REST API response.
+	 * @param int                         $status   Optional. Status code.
+	 * @param string                      $message  Optional. Message to display when the assertion fails.
 	 */
 	protected function assertErrorResponse( $code, $response, $status = null, $message = '' ) {
 
