diff --git a/includes/sqlite-database-integration/load.php b/includes/sqlite-database-integration/load.php
index 3af80903..ddc5a9b6 100644
--- a/includes/sqlite-database-integration/load.php
+++ b/includes/sqlite-database-integration/load.php
@@ -12,7 +12,9 @@
  * @package wp-sqlite-integration
  */
 
-define( 'SQLITE_MAIN_FILE', __FILE__ );
+if (!defined('SQLITE_MAIN_FILE')) {
+    define('SQLITE_MAIN_FILE', __FILE__);
+}
 
 require_once __DIR__ . '/admin-page.php';
 require_once __DIR__ . '/activate.php';
